File: 2024/kt/src/main/kotlin/Day5.kt
package org.example

import java.io.File

fun day5() {
    val input = File("inputs/input5").readLines()
    val orders = mutableSetOf<Pair<Int, Int>>()
    val pages = mutableListOf<List<Int>>()

    // Parse input
    input.forEach { line ->
        when {
            line.contains("|") -> {
                val (a, b) = line.split("|")
                orders.add(Pair(a.toInt(), b.toInt()))
            }
            line.contains(",") -> {
                pages.add(line.split(",").map { it.toInt() })
            }
        }
    }

    val rightOrders = pages.filter { page ->
        page.indices.all { index ->
            val current = page[index]
            val rest = page.subList(index + 1, page.size)
            rest.all { next ->
                orders.contains(Pair(current, next))
            }
        }
    }

    val wrongOrders = pages.filter { page -> !rightOrders.contains(page) }

    val mids = rightOrders.mapNotNull { it.middle() }
    println(mids.sum()) // This is Part 1

    val newRightOrders = mutableListOf<List<Int>>()

    wrongOrders.forEach { ord ->
        val perms = ord.lazyPermutations()
        for (p in perms) {
            if (p.indices.all { index ->
                    val current = p[index]
                    val rest = p.subList(index + 1, p.size)
                    rest.all { next ->
                        orders.contains(Pair(current, next))
                    }
                }) {
                newRightOrders.add(p)
                break
            }
        }
    }

    val mids2 = newRightOrders.mapNotNull { it.middle() }
    println(mids2.sum()) // This is for part 2
}

fun <T> List<T>.middle(): T? {
    return if (this.isNotEmpty()) this[this.size / 2] else null
}

fun <T : Number> List<T>.sum(): Int {
    return this.fold(0) { acc, num -> acc + num.toInt() }
}

fun <T> List<T>.lazyPermutations(): Sequence<List<T>> = sequence {
    if (this@lazyPermutations.size == 1) {
        yield(this@lazyPermutations)
    } else {
        for (i in this@lazyPermutations.indices) {
            val element = this@lazyPermutations[i]
            val remaining = this@lazyPermutations.filterIndexed { index, _ -> index != i }
            for (perm in remaining.lazyPermutations()) {
                yield(listOf(element) + perm)
            }
        }
    }
}

File: 2024/kt/src/main/kotlin/Day5.py
def parse_input(file_path):
    orders = set()
    pages = []

    with open(file_path, "r") as file:
        lines = file.readlines()

    for line in lines:
        line = line.strip()
        if "|" in line:
            a, b = map(int, line.split("|"))
            orders.add((a, b))
        elif "," in line:
            pages.append(list(map(int, line.split(","))))

    return orders, pages


def middle(lst):
    """Return the middle element of a list."""
    return lst[len(lst) // 2] if lst else None


def sum_list(lst):
    """Sum elements of a list."""
    return sum(lst)


def find_valid_permutation(order, orders):
    """Find the first valid permutation of 'order' that satisfies 'orders'."""

    def backtrack(path, remaining):
        if not remaining:
            # Check if the constructed path is valid
            if all(
                all((path[i], path[j]) in orders for j in range(i + 1, len(path)))
                for i in range(len(path))
            ):
                return path
            return None

        for i, item in enumerate(remaining):
            result = backtrack(path + [item], remaining[:i] + remaining[i + 1 :])
            if result:
                return result

        return None

    return backtrack([], order)


def day5():
    file_path = "./../../../inputs/input5"
    orders, pages = parse_input(file_path)

    right_orders = []
    wrong_orders = []

    # Classify pages as right or wrong
    for page in pages:
        if all(
            all((page[i], page[j]) in orders for j in range(i + 1, len(page)))
            for i in range(len(page))
        ):
            right_orders.append(page)
        else:
            wrong_orders.append(page)

    # Part 1: Middle element sum of right orders
    mids = [middle(page) for page in right_orders if middle(page) is not None]
    print(sum_list(mids))

    # Part 2: Fix wrong orders
    new_right_orders = []

    for ord in wrong_orders:
        valid_perm = find_valid_permutation(ord, orders)
        if valid_perm:
            new_right_orders.append(valid_perm)

    mids2 = [middle(page) for page in new_right_orders if middle(page) is not None]
    print(sum_list(mids2))


if __name__ == "__main__":
    day5()
